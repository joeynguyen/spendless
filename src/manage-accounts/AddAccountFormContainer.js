import React, { Component, PropTypes } from 'react';
import { bindActionCreators } from 'redux';
import toastr from 'toastr';
import AddAccountForm from './AddAccountForm.js';
import { reduxForm } from 'redux-form';
import * as accountsActions from '../account/AccountsActions.js';

export class AddAccountFormContainer extends Component {
  static propTypes = {
    actions: PropTypes.object.isRequired,
    fields: PropTypes.object.isRequired,
    handleSubmit: PropTypes.func.isRequired,
    resetForm: PropTypes.func.isRequired,
    submitting: PropTypes.bool.isRequired,
    visible: PropTypes.bool.isRequired,
  }

  componentDidUpdate(prevProps) {
    // reset form forms on clicking Cancel button
    if (this.props.visible !== prevProps.visible && this.props.visible === false) {
      // wait 200ms for slideUp animation to finish so transition looks nicer
      setTimeout(() => {
        this.props.resetForm();
      }, 200);
    }
  }
  localHandleSubmit = () => {
    const newAccount = {
      '_id': new Date().toISOString(),
      'name': this.props.fields.accountName.value,
      'type': this.props.fields.accountType.value,
      'company': this.props.fields.accountCompany.value,
    };

    this.props.actions.saveAccount(newAccount)
      .then(result => {
        toastr.success(result.name + ' account added', null, {timeOut: 1500});
        // remove focus from input field that it was on when form was submitted
        // if we hit the Enter key to submit the form the focus will stay on the
        // input and show an error on that input because it will be marked as 'touched'
        // by redux-form when we move the focus to another field, not good for accountCompany
        // field because it's hidden and will display with an error when shown
        document.activeElement.blur();
        // Reset AddAccount form fields
        this.props.resetForm();
      })
      .catch(() => {
        toastr.error('Restart the application and retry', 'Error adding account', {timeOut: 1500});
      });
  }
  render() {
    const reduxFormHandleSubmit = this.props.handleSubmit(this.localHandleSubmit);
    return (
      <AddAccountForm
        fields={this.props.fields}
        localHandleSubmit={reduxFormHandleSubmit}
        submitting={this.props.submitting}
      />
    );
  }
}

function validateForm(values) {
  const errors = {};

  if (!values.accountName) {
    errors.accountName = 'Enter an account name';
  }
  if (!values.accountType) {
    errors.accountType = 'Enter an account type';
  }
  if (!values.accountCompany) {
    errors.accountCompany = 'Enter the name of the financial institution';
  }

  return errors;
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(accountsActions, dispatch)
  };
}

// connect: 1st argument is mapStateToProps, 2nd state is mapDispatchToProps
// reduxForm: 1st argument is form config, 2nd is mapStateToProps, 3rd state is mapDispatchToProps
// initialValues required for resetForm
export default reduxForm(
  {
    form: 'AddAccount',
    fields: ['accountName', 'accountType', 'accountCompany'],
    initialValues: {
      accountName: '',
      accountType: '',
      accountCompany: '',
    },
    validate: validateForm
  },
  null,
  mapDispatchToProps
)(AddAccountFormContainer);
