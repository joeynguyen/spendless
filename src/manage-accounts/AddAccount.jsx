import React, { Component } from 'react';
import { AddIcon, Animate, Box, Button } from 'grommet';
import AddAccountFormContainer from './AddAccountFormContainer.jsx';

export default class AddAccount extends Component {
  state = {
    addAccountVisible: false,
  }
  toggleAddAccount = () => {
    this.setState({ addAccountVisible: !this.state.addAccountVisible });
  }
  render() {
    const addAccountVisible = this.state.addAccountVisible;
    const btnIcon = addAccountVisible ? null : <AddIcon />;
    const btnLabel = addAccountVisible ? 'Cancel' : 'Add Account';
    return (
      <div>
        <Box size="small" pad={{vertical: 'medium'}}>
          <Button
            icon={btnIcon}
            critical={addAccountVisible}
            label={btnLabel}
            onClick={this.toggleAddAccount}
          />
        </Box>
        <Animate
          visible={addAccountVisible}
          enter={{animation: 'slide-down', duration: 300, delay: 0}}
        >
          <AddAccountFormContainer visible={this.state.addAccountVisible} />
        </Animate>
      </div>
    );
  }
}
