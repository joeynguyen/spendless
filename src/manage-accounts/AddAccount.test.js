import expect from 'expect';
import React from 'react';
import { shallow } from 'enzyme';
import AddAccount from './AddAccount';
import AddAccountFormContainer from './AddAccountFormContainer.js';

function setup() {
  return shallow(<AddAccount />);
}

describe('AddAccount', () => {
  describe('by default', () => {
    const component = setup();

    it('has a Button', (done) => {
      expect(component.find('Button').length).toBe(1);
      done();
    });

    it('has a Panel', (done) => {
      expect(component.find('Panel').length).toBe(1);
      done();
    });

    it('has an AddAccountFormContainer', (done) => {
      expect(component.find(AddAccountFormContainer).length).toBe(1);
      done();
    });

    it('addAccountVisible state is false by default', (done) => {
      expect(component.state().addAccountVisible).toBe(false);
      done();
    });
  });

  describe('when Button is clicked', () => {
    const component = setup();

    expect(component.state().addAccountVisible).toBe(false);
    expect(component.find('Panel').props().expanded).toBe(false);
    component.find('Button').simulate('click');

    it('addAccountVisible state to be true when Button is clicked', (done) => {
      expect(component.state().addAccountVisible).toBe(true);
      done();
    });

    it('form section is hidden by default but shows when Button is clicked', (done) => {
      expect(component.find('Panel').props().expanded).toBe(true);
      done();
    });
  });
});
