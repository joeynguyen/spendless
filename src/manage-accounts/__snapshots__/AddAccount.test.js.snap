// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAccount by default Add Account button is shown and form section is hidden 1`] = `
<div
  style={
    Object {
      "marginBottom": "15px",
    }
  }
>
  <Button
    block={false}
    ghost={false}
    icon="plus"
    loading={false}
    onClick={[Function]}
    prefixCls="ant-btn"
    size="large"
    type="primary"
  >
     Add Account
  </Button>
  <Collapse
    fixedHeight={-1}
    forceInitialAnimation={false}
    hasNestedCollapse={false}
    isOpened={false}
    onMeasure={[Function]}
    onRender={[Function]}
    onRest={[Function]}
    style={Object {}}
    theme={
      Object {
        "collapse": "ReactCollapse--collapse",
        "content": "ReactCollapse--content",
      }
    }
  >
    <Form(AddAccountForm) />
  </Collapse>
</div>
`;

exports[`AddAccount when Button is clicked form section is shown and button changes to Cancel 1`] = `
<div
  style={
    Object {
      "marginBottom": "15px",
    }
  }
>
  <Button
    block={false}
    ghost={false}
    icon=""
    loading={false}
    onClick={[Function]}
    prefixCls="ant-btn"
    size="large"
    type="danger"
  >
    Cancel
  </Button>
  <Collapse
    fixedHeight={-1}
    forceInitialAnimation={false}
    hasNestedCollapse={false}
    isOpened={true}
    onMeasure={[Function]}
    onRender={[Function]}
    onRest={[Function]}
    style={Object {}}
    theme={
      Object {
        "collapse": "ReactCollapse--collapse",
        "content": "ReactCollapse--content",
      }
    }
  >
    <Form(AddAccountForm) />
  </Collapse>
</div>
`;
